src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -152,6 +154,88 @@ private MethodBase FunctionLookup(IList args)
             return mi;
         }
 
+        private static HashSet<Type> GetTypeStatisticsForArray(ExprListNode array)
+        {
+            var arrayTypes = new HashSet<Type>();
+
+            foreach (var exp in array.Exprs)
+            {
+                if (exp is ExprListNode eln)
+                {
+                    var subArray = GetTypeStatisticsForArray(eln);
+                    var t = GetOverallTypeForArray(subArray);
+
+                    if (t != typeof(object))
+                    {
+                        t = t.MakeArrayType();
+                    }
+                    arrayTypes.Add(t);
+                }
+                else
+                {
+                    Type t;
+                    switch (exp.Kind)
+                    {
+                        case AstKind.Integer:
+                            t = typeof(long);
+                            break;
+                        case AstKind.Double:
+                            t = typeof(double);
+                            break;
+                        case AstKind.Boolean:
+                            t = typeof(bool);
+                            break;
+                        case AstKind.Char:
+                            t = typeof(char);
+                            break;
+                        case AstKind.String:
+                            t = typeof(string);
+                            break;
+                        case AstKind.Identifier:
+                            t = typeof(object);
+                            break;
+                        default:
+                            t = typeof(object);
+                            break;
+                    }
+                    arrayTypes.Add(t);
+                }
+            }
+            return arrayTypes;
+        }
+
+        private static Type GetOverallTypeForArray(HashSet<Type> arrayTypes)
+        {
+            if (arrayTypes.Count == 1)
+            {
+                // There is only a single type in the array, return it.
+                return arrayTypes.FirstOrDefault();
+            }
+            if(arrayTypes.Count == 2)
+            {
+                bool isLong = false;
+                bool isDouble = false;
+                bool isChar = false;
+                bool isString = false;
+                foreach (var type in arrayTypes)
```

Comment:
if the length of the array is only 2...I guess my suggestion will not improve performance at all.

Reply:
I initially though we were iterating the entire array (the items not the types).